lib/Common/DataStructures/ClusterList.h

```diff
@@ -0,0 +1,402 @@
+            backingStore = tempList;
+            AllocatorDeleteArray(TAllocator, alloc, curNumSegments, oldBackingStore);
+        }
+    }
+    inline void CreateBacking(indexType index)
+    {
+        Assert(index >= maxIndex || backingStore[index / numPerSegment] == nullptr);
+        // grow the size of the pointer array if needed
+        EnsureBaseSize(index + 1);
+        if (backingStore[index / numPerSegment] == nullptr)
+        {
+            // allocate a new segment
+            backingStore[index / numPerSegment] = AllocatorNewArrayLeaf(TAllocator, alloc, indexType, numPerSegment);
+            indexType baseForSegment = (index / numPerSegment) * numPerSegment;
+            for (indexType i = 0; i < numPerSegment; i++)
+            {
+                backingStore[index / numPerSegment][i] = i + baseForSegment;
+            }
+        }
+    }
+    template<bool createBacking = false, bool assumeConsolidated = false>
+    inline indexType Lookup(indexType index)
+    {
+        if (index >= maxIndex || backingStore[index / numPerSegment] == nullptr)
+        {
+            // Not being there simply means that it's still an identity mapping
+            if(createBacking)
+            {
+                CreateBacking(index);
+            }
+            return index;
+        }
+        if (assumeConsolidated)
+        {
+            // If the list is consolidated, then we can return whatever's there
+            return backingStore[index / numPerSegment][index % numPerSegment];
+        }
+        else
+        {
+            // If it's not consolidated, we need to follow the chain down, and update each entry to the root
+            indexType currentValue = backingStore[index / numPerSegment][index % numPerSegment];
+            if (currentValue == index)
+            {
+                return index;
+            }
+            indexType trueRoot = Lookup<false, assumeConsolidated>(currentValue);
+            backingStore[index / numPerSegment][index % numPerSegment] = trueRoot;
+            return trueRoot;
+        }
+    }
+
+    template<bool assumeExists = false>
+    inline void Assign(indexType index, indexType value)
+    {
+        if (!assumeExists)
+        {
+            if (index >= maxIndex || backingStore[index / numPerSegment] == nullptr)
+            {
+                CreateBacking(index);
+            }
+        }
+        backingStore[index / numPerSegment][index % numPerSegment] = value;
+    }
+public:
+    // Create a new ClusterList with a specified maximum node number
+    SegmentClusterList(TAllocator* allocator, indexType maxCount)
+        : maxIndex(maxCount % numPerSegment == 0 ? maxCount : maxCount + (numPerSegment - (maxCount % numPerSegment)))
+        , backingStore(nullptr)
+        , alloc(allocator)
+#if DBG
+        , consolidated(true)
+#endif
+    {
+        backingStore = (indexType**)AllocatorNewArrayZ(TAllocator, alloc, indexType*, maxIndex / numPerSegment);
```

Comment:
Why is this cast necessary? It looks to me like `AllocateArray` already returns `T*`.